### ISSUE TITLE
Placing model on bfloat16 on CPU make it freeze/hang

### ISSUE THREAD

**Issue by Ierezell** [pytorch/pytorch#75458]
### 🐛 Describe the bug

Note : 
I know that bfloat16 should obviously not be used on a CPU model.
Maybe it's a better practice to do `to(self.device).to(bfloat16)` than `.to(bfloat16).to(self.device)`
But there will be dummy guys like me who will also hit the same error. 

Maybe instead of hanging forever we could have an Exception or some warning.
 
```python
my_big_model = load_my_model()  # Model is on CPU
my_big_model.to(torch.bfloat16)

my_big_model.inference(my_input)
# Hang forever
```

```python
my_big_model = load_my_model()  # Model is on CPU
my_big_model.inference(my_input)
# Works perfectly
```

```python
my_big_model = load_my_model()  # Model is on CPU
if running_on_gpu:
    my_big_model.model.to(bfloat16).to(self.device)
my_big_model.inference(my_input)
# Run perfectly
```

Thanks, 
Have a great day :)

### Versions

PyTorch version: 1.11.0+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.10 (main, Jan 15 2022, 18:17:56)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.13.0-37-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 470.103.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.2
[pip3] pytorch-pretrained-bert==0.6.2
[pip3] torch==1.11.0+cu113
[conda] Could not collect

**Comment by vadimkantorov** [comment]
(unrelated to the hang, but can one fuse `to(self.device).to(bfloat16)` into `to(device = self.device, dtype = torch.bfloat16)`?)

**Comment by Ierezell** [comment]
Hi @vadimkantorov, thanks !
This will become my new go to. 
It's always a problem to know if it's better to go first on GPU than to cast or the other way. using your solution will let pytorch do the best :) (maybe it should be emphasized on the docs ?)

**Comment by leslie-fang-intel** [comment]
Hi @Ierezell, 

- Bfloat16 can be used on a CPU model (even so not all bf16 cpu optimization has been landed yet). Suggested to use [Autocast](https://pytorch.org/docs/master/amp.html#torch.autocast) for BFloat16 on CPU.
- Another thing, these code `model.to(torch.bfloat16)` should not hang on CPU even without Autocast.
  - Is your model a convolution based model? The good bf16 convolution performance depends on this [PR] (https://github.com/pytorch/pytorch/pull/55864). Without this PR, in my system, the BF16 convolution is very slow.
  - Also you can get further BF16 HW speed-up on CPU codename of CPX and future SPR.
 ```
my_big_model = load_my_model()  # Model is on CPU
my_big_model.to(torch.bfloat16)
my_big_model.inference(my_input) # Hang forever
```

**Comment by Ierezell** [comment]
Hi @leslie-fang-intel, thanks for the reply! 
Good to know that some bf16 optimizations are coming to CPU.

No, my model is a hugging face transformer one (maybe I should raise the issue there ?). 

I also thought about it being slow but it was a matter of seconds (without bf16 or on GPU) and with bf16 on CPU it just computed for more than 30mn.... (also none of the processors where running (checked with htop) vs when not using bf16 processors where all 100%)

 I hope you will be able to reproduce the error, as I said adding a one-line check resolved the problem for me but I guess it is worth investigating that. 

Have a great day

**Comment by leslie-fang-intel** [comment]
Hi @Ierezell, thanks for the feedback. 
@liangan1 can help looking into the hugging face transformer model's performance. Could you provide more details' information? 

- Please share your hugging face transformer model repo and steps to reproduce.
- Please share which pytorch version are you using.
- Please share which CPU platform are you using.

>  as I said adding a one-line check resolved the problem for me but I guess it is worth investigating that.

Did you mean add one-line check then you can run your bf16 model quickly? What's the line did you add?

**Comment by Ierezell** [comment]
Hi @leslie-fang-intel,

I got the problem with `allenai/macaw-3b` (I know it's a bit borderline and it's a huge model)

Here is a snippet : 

```python
# Device : device = torch.device("cuda:0") if torch.cuda.is_available()  else torch.device("cpu")
tokenizer: PreTrainedTokenizer = AutoTokenizer.from_pretrained("allenai/macaw-3b")
model: PreTrainedModel = AutoModelForSeq2SeqLM.from_pretrained("allenai/macaw-3b")

if "cuda" in self.device:  # Without this line (I mean always placing model on CPU I get the error as described in the first post)
    model.to(bfloat16).to(device)
  

input_string = some_string_preprocessing() # As showed here : https://github.com/allenai/macaw/blob/5a0b32b7879e39cb43551830833dcaed8e208773/macaw/utils.py
# I mostly copy pasted their util code to create the right string. However it can be hardcoded as they show examples in the readme. 

with no_grad():
            input_ids = self.tokenizer.batch_encode_plus(
                input_strings, return_tensors="pt",
                add_special_tokens=True,
                padding=True,
                truncation=True
            )
            input_ids = input_ids.to(self.device)

            res = model.generate(
                    input_ids["input_ids"],
                    attention_mask=input_ids["attention_mask"],
                    **generator_options
                ),
            )
            output_strings = self.tokenizer.batch_decode(res, skip_special_tokens=True)
```

With this piece of code, without bf16, it runs on CPU and GPU. 
When using Bfloat16 on GPU it worked fine but when using bfloat16 (by mistake) on CPU then it "froze". 

Note that on CPU without bf16 I was able to get the inference in less than 30s but I waited more than 20mn for CPU with bf16...

Thanks for your help, 
Have a great day

**Comment by leslie-fang-intel** [comment]
Thanks @Ierezell. From the snippet of code, we didn't see obviously error. 
Since we suspect there are some BF16 kernels which are extreme slow, could you kindly provide more information for us to reproduce:
1. Model repo and steps to run the model, then we can profiling the model performance.
2. The pytorch version and CPU information(use command of `lscpu` may be the first step to check cpu information).

**Comment by Ierezell** [comment]
Hi @leslie-fang-intel, 

It's a private repo of the company I'm working for... However, this code was used to compute a paraphrasing metric to evaluate text as follow : 

```python
from transformers.pipelines.zero_shot_classification import ZeroShotClassificationPipeline
from transformers.pipelines import pipeline
from typing_extensions import TypedDict
from typing import List
from abc import ABC, abstractmethod

class HuggingFaceZeroShotOutput(TypedDict):
    sequence: str
    labels: List[str]
    scores: List[float]

class Metric(ABC):
    @abstractmethod
    def __call__(self, prediction: str, reference: str) -> float:
        ...

class ParaphrasingMetric(Metric):
    def __init__(self) -> None:
        self.nli = cast(
            ZeroShotClassificationPipeline,
            pipeline(
                "zero-shot-classification",
                model="sentence-transformers/paraphrase-MiniLM-L6-v2",
                tokenizer="sentence-transformers/paraphrase-MiniLM-L6-v2",
            ),
        )

    def __call__(self, prediction: str, reference: str) -> float:
        res: HuggingFaceZeroShotOutput = self.nli(reference, prediction, max_length=512)
        score: float = res["scores"][0]
        return score
```

Then you can call it with any pair of strings.

My OS infos : 

LsCpu
```
Architecture :                              x86_64
  Mode(s) opératoire(s) des processeurs :   32-bit, 64-bit
  Tailles des adresses:                     39 bits physical, 48 bits virtual
  Boutisme :                                Little Endian
Processeur(s) :                             12
  Liste de processeur(s) en ligne :         0-11
Identifiant constructeur :                  GenuineIntel
  Nom de modèle :                           Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
    Famille de processeur :                 6
    Modèle :                                165
    Thread(s) par cœur :                    2
    Cœur(s) par socket :                    6
    Socket(s) :                             1
    Révision :                              2
    multiplication des MHz du/des CPU(s) :  44%
    Vitesse maximale du processeur en MHz : 5000,0000
    Vitesse minimale du processeur en MHz : 800,0000
    BogoMIPS :                              5202.65
    Drapeaux :                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 c
                                            lflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp 
                                            lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop
                                            _tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 sss
                                            e3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_dea
                                            dline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fau
                                            lt epb invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi 
                                            flexpriority ept vpid ept_ad fsgsbase tsc_adjust sgx bmi1 avx2 smep bmi2 
                                            erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xget
                                            bv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp 
                                            pku ospke sgx_lc md_clear flush_l1d arch_capabilities
Fonctionnalités de virtualisation :         
  Virtualisation :                          VT-x
Caches (somme de toutes) :                  
  L1d :                                     192 KiB (6 instances)
  L1i :                                     192 KiB (6 instances)
  L2 :                                      1,5 MiB (6 instances)
  L3 :                                      12 MiB (1 instance)
NUMA :                                      
  Nœud(s) NUMA :                            1
  Nœud NUMA 0 de processeur(s) :            0-11
Vulnérabilités :                            
  Itlb multihit :                           KVM: Mitigation: VMX disabled
  L1tf :                                    Not affected
  Mds :                                     Not affected
  Meltdown :                                Not affected
  Spec store bypass :                       Mitigation; Speculative Store Bypass disabled via prctl
  Spectre v1 :                              Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2 :                              Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
  Srbds :                                   Not affected
  Tsx async abort :                         Not affected
```
(It's in french but you have important infos, sorry for that)

I'm running ArchLinux (normal kernel 5.17.3-arch1-1)

Don't hesitate if you need more infos.

**Comment by leslie-fang-intel** [comment]
Hi @Ierezell,
Thanks for providing the general use case. It still looks like not runnable for us to profile and identify the performance bottleneck of CPU BF16 case for this model.
I remember you said your private model is based on `allenai/macaw-3b`. I believe it's based on this public repo(https://github.com/allenai/macaw). Will you see similar problem with this public repo or it's specific for your private model?

**Comment by Ierezell** [comment]
Hi @leslie-fang-intel, 

Yes it's this repo, they have some demo code I cleaned a bit for my usecase. 
Also, this model is available on HuggingFace directly so it's easy to use if you want to try on your side. 

I will try the oneDNN flag directly on their repo to tell you if I have the same performance bottleneck.

From their repo: first example
```python
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
tokenizer = AutoTokenizer.from_pretrained("allenai/macaw-3b") # you can use allenai/macaw-large which is smaller
model = AutoModelForSeq2SeqLM.from_pretrained("allenai/macaw-3b")
input_string = "$answer$ ; $mcoptions$ ; $question$ = What is the color of a cloudy sky?"
input_ids = tokenizer.encode(input_string, return_tensors="pt")
output = model.generate(input_ids, max_length=200)
```

I'm good with the solution of a small if condition that fixes my problem but I'm happy to help you pinpoint this problem for other that might ran in it.

**Comment by Dobbie123** [comment]
This here me !$&@90in-cuinPrIcoon!!Zoonxihafey/)
Bfiitchs☠️☠️☠️💩👺👹😹👀😎✌️👌👇🤌🫴🤏$$$$✍️🖕🖕🖕🖕🫵💋👁👤
